// play-dl/dist/Spotify/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.refreshToken = exports.sp_search = exports.is_expired = exports.SpotifyAuthorize = exports.sp_validate = exports.spotify = void 0;
const request_1 = require("../YouTube/utils/request");
const classes_1 = require("./classes");
const fs_1 = __importDefault(require("fs"));
let spotifyData;
if (fs_1.default.existsSync('.data/spotify.data')) {
    spotifyData = JSON.parse(fs_1.default.readFileSync('.data/spotify.data').toString());
}
const pattern = /^((https:)?\/\/)?open.spotify.com\/(track|album|playlist)\//;
async function spotify(url) {
    if (!spotifyData)
        throw new Error('Spotify Data is missing\nDid you forgot to do authorization ?');
    if (!url.match(pattern))
        throw new Error('This is not a Spotify URL');
    if (url.indexOf('track/') !== -1) {
        const trackID = url.split('track/')[1].split('&')[0].split('?')[0];
        const response = await request_1.request(`https://api.spotify.com/v1/tracks/${trackID}?market=${spotifyData.market}`, {
            headers: {
                Authorization: `${spotifyData.token_type} ${spotifyData.access_token}`
            }
        }).catch((err) => {
            return err;
        });
        if (response instanceof Error)
            throw response;
        return new classes_1.SpotifyVideo(JSON.parse(response));
    }
    else if (url.indexOf('album/') !== -1) {
        const albumID = url.split('album/')[1].split('&')[0].split('?')[0];
        const response = await request_1.request(`https://api.spotify.com/v1/albums/${albumID}?market=${spotifyData.market}`, {
            headers: {
                Authorization: `${spotifyData.token_type} ${spotifyData.access_token}`
            }
        }).catch((err) => {
            return err;
        });
        if (response instanceof Error)
            throw response;
        return new classes_1.SpotifyAlbum(JSON.parse(response), spotifyData);
    }
    else if (url.indexOf('playlist/') !== -1) {
        const playlistID = url.split('playlist/')[1].split('&')[0].split('?')[0];
        const response = await request_1.request(`https://api.spotify.com/v1/playlists/${playlistID}?market=${spotifyData.market}`, {
            headers: {
                Authorization: `${spotifyData.token_type} ${spotifyData.access_token}`
            }
        }).catch((err) => {
            return err;
        });
        if (response instanceof Error)
            throw response;
        return new classes_1.SpotifyPlaylist(JSON.parse(response), spotifyData);
    }
    else
        throw new Error('URL is out of scope for play-dl.');
}
exports.spotify = spotify;
function sp_validate(url) {
    if (!url.match(pattern))
        return false;
    if (url.indexOf('track/') !== -1) {
        return 'track';
    }
    else if (url.indexOf('album/') !== -1) {
        return 'album';
    }
    else if (url.indexOf('playlist/') !== -1) {
        return 'playlist';
    }
    else
        return false;
}
exports.sp_validate = sp_validate;
async function SpotifyAuthorize(data) {
    const response = await request_1.request(`https://accounts.spotify.com/api/token`, {
        headers: {
            'Authorization': `Basic ${Buffer.from(`${data.client_id}:${data.client_secret}`).toString('base64')}`,
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: `grant_type=authorization_code&code=${data.authorization_code}&redirect_uri=${encodeURI(data.redirect_url)}`,
        method: 'POST'
    }).catch((err) => {
        return err;
    });
    if (response instanceof Error)
        throw response;
    const resp_json = JSON.parse(response);
    spotifyData = {
        client_id: data.client_id,
        client_secret: data.client_secret,
        redirect_url: data.redirect_url,
        access_token: resp_json.access_token,
        refresh_token: resp_json.refresh_token,
        expires_in: Number(resp_json.expires_in),
        expiry: Date.now() + (resp_json.expires_in - 1) * 1000,
        token_type: resp_json.token_type,
        market: data.market
    };
    fs_1.default.writeFileSync('.data/spotify.data', JSON.stringify(spotifyData, undefined, 4));
    return true;
}
exports.SpotifyAuthorize = SpotifyAuthorize;
function is_expired() {
    if (Date.now() >= spotifyData.expiry)
        return true;
    else
        return false;
}
exports.is_expired = is_expired;
async function sp_search(query, type, limit = 10) {
    const results = [];
    if (!spotifyData)
        throw new Error('Spotify Data is missing\nDid you forgot to do authorization ?');
    if (query.length === 0)
        throw new Error('Pass some query to search.');
    if (limit > 50 || limit < 0)
        throw new Error(`You crossed limit range of Spotify [ 0 - 50 ]`);
    const response = await request_1.request(`https://api.spotify.com/v1/search?type=${type}&q=${query.replaceAll(' ', '+')}&limit=${limit}&market=${spotifyData.market}`, {
        headers: {
            Authorization: `${spotifyData.token_type} ${spotifyData.access_token}`
        }
    }).catch((err) => {
        return err;
    });
    if (response instanceof Error)
        throw response;
    const json_data = JSON.parse(response);
    if (type === 'track') {
        json_data.tracks.items.forEach((track) => {
            results.push(new classes_1.SpotifyVideo(track));
        });
    }
    else if (type === 'album') {
        json_data.albums.items.forEach((album) => {
            results.push(new classes_1.SpotifyAlbum(album, spotifyData));
        });
    }
    else if (type === 'playlist') {
        json_data.playlists.items.forEach((playlist) => {
            results.push(new classes_1.SpotifyPlaylist(playlist, spotifyData));
        });
    }
    return results;
}
exports.sp_search = sp_search;
async function refreshToken() {
    const response = await request_1.request(`https://accounts.spotify.com/api/token`, {
        headers: {
            'Authorization': `Basic ${Buffer.from(`${spotifyData.client_id}:${spotifyData.client_secret}`).toString('base64')}`,
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: `grant_type=refresh_token&refresh_token=${spotifyData.refresh_token}`,
        method: 'POST'
    }).catch((err) => {
        return err;
    });
    if (response instanceof Error)
        return false;
    const resp_json = JSON.parse(response);
    spotifyData.access_token = resp_json.access_token;
    spotifyData.expires_in = Number(resp_json.expires_in);
    spotifyData.expiry = Date.now() + (resp_json.expires_in - 1) * 1000;
    spotifyData.token_type = resp_json.token_type;
    fs_1.default.writeFileSync('.data/spotify.data', JSON.stringify(spotifyData, undefined, 4));
    return true;
}
exports.refreshToken = refreshToken;
